### ISSUE TITLE
The frame had a length of 272035850, but we only support messages up to 267386880 bytes long

### ISSUE THREAD

**Issue by vijayprakashjoshi** [yugabyte/yugabyte-db#4679]
Jira Link: [DB-2556](https://yugabyte.atlassian.net/browse/DB-2556)
We have a table in which we have stored large amount of data.
When we are querying that table and performing operations on it, we see error saying we exceeded frame size.

Query:
Our query somewhat looks like this:
`SELECT r->>1 as label, avg(cast(r->>8 as bigint)) as AvgTime FROM <Table-Name> tbl, json_array_elements(cast(tbl.jtl as json)) r group by r->>1 ;`

Error:
`Error Timed out: Read(tablet: 652210427f6a490fa04ea983e321ee7c, num_ops: 1, num_attempts: 184, txn: 00000000-0000-0000-0000-000000000000) passed its deadline 48890.078s (passed: 60.176s): Network error (yb/rpc/binary_call_parser.cc:94): The frame had a length of 272035850, but we only support messages up to 267386880 bytes long.`

**Comment by kmuthukk** [comment]
hi @vijayprakashjoshi --

Is reporting/OLAP style queries your more common use case? 

While YugabyteDB is not optimized for these full-table scan/reporting type workloads, do agree that we shouldn't be getting this error, and instead be paging the requests internally into appropriate / reasonable sized chunks.

Few questions:
* approx how many rows in the table?
* and how big is the typical row size or the json payload size on average (as well as max)?

**Comment by kmuthukk** [comment]
And wondering if reducing the yb-tserver gflag `--ysql_prefetch_limit=1024` to a lower value would be a temporary workaround to this issue.

CC: @ndeodhar , @m-iancu .

**Comment by vijayprakashjoshi** [comment]
> hi @vijayprakashjoshi --
> 
> Is reporting/OLAP style queries your more common use case?
> 
> While YugabyteDB is not optimized for these full-table scan/reporting type workloads, do agree that we shouldn't be getting this error, and instead be paging the requests internally into appropriate / reasonable sized chunks.
> 
> Few questions:
> 
> * approx how many rows in the table?
@kmuthukk We have only 327 rows.
select count(*) from <Table_Name>;                                                                                                                              count
-------
   327
(1 row)

> * and how big is the typical row size or the json payload size on average (as well as max)?
Each row would be ~ 9.4 MB.

**Comment by kmuthukk** [comment]
Ok with payload size of ~9MB for the row, and 1024 prefetch size -- that could explain the issue. 

Internally, we need to automatically reduce the prefetching size if the payload is above the RPC thresholds.

**Comment by vijayprakashjoshi** [comment]
@kmuthukk 
We increased the `--rpc_max_message_size=300` in `/home/ec2-user/yugabyte-db/tserver/conf/server.conf` and restarted yb-tservers.

Can you please suggest if that would be correct approach? Or should we reduce this `--ysql_prefetch_limit=1024` ?

**Comment by kmuthukk** [comment]
Increasing the rpc_max_message_size to 300MB is not the right workaround for this issue IMO. You might still hit the same problem at larger table sizes. As the number of the rows in the table increases, the rows in each shard will also increase. And trying to prefetch 1024 of these rows can still end up putting you over the rpc_max_message_size limit. So my guess (which should confirm) is that this is not the correct fix.

The other fix of reducing the prefetch size to a lower value (say 8 or 16) will probably serve better in the short-run. Nevertheless, that fix is sub-optimal because for other tables where the rows are smaller you do want to prefetch more aggressively.

So the right fix might really be for YugabyteDB to automatically reduce the internal YSQL prefetch size when payload being processed internally goes above a certain threshold.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4679-analysis.md`):

# Analysis: yugabyte/yugabyte-db#4679 (DB-2556)

Issue: "The frame had a length of 272035850, but we only support messages up to 267386880 bytes long"
(opened 2020-06-04, area/ysql, kind/bug, community/request; 7 comments, last 2022-09-20). Reporter: a table of
327 rows, each ~9.4 MB of JSON text, and a `json_array_elements(...)` GROUP BY over the whole table; after 184
retries the statement fails with a read RPC that "passed its deadline" because the response frame
(272,035,850 bytes) exceeded `rpc_max_message_size` (255 MiB = 267,386,880). Maintainer comments agree the
engine "should be paging the requests internally into appropriate / reasonable sized chunks" and point at the
row-count prefetch (`ysql_prefetch_limit=1024`). Analysis date: 2026-09-06.

## Verdict
NOT-REPRODUCED - the row-count-only paging that produced the oversized frame is gone from every current
in-support release and the latest preview (deterministic run on 2024.2.11.0-b36, 2025.1.4.0-b103, 2025.2.6.0-b111,
2026.1.1.1-b2, 2.25.2.0-b359: the reporter-shaped query and a plain full scan of ~135 MB of wide rows succeed with
default GUCs under an 8 MiB frame limit), while the same harness on the era release 2.1.6.0-b17 reproduces the
exact reported signature (positive control). Fixed by the DocDB response-size cap introduced in 2025 (#24627,
#26375, #27202); the only residual is that a single value larger than `rpc_max_message_size` cannot be stored or
read, which is the frame limit itself, not a paging defect.

## Root cause
Era mechanism (v2.1.6, the release family of the report):
- pggate pages sequential scans by row count only: `src/yb/yql/pggate/pggate_flags.cc:51`
  (`ysql_prefetch_limit=1024`) feeds `src/yb/yql/pggate/pg_doc_op.cc:106,295,314` (`req->set_limit(...)`); no
  size limit exists in the request.
- DocDB fills one response until the row limit or end of tablet: `src/yb/docdb/pgsql_operation.cc:434-439,484`
  (`row_count_limit` is the only bound). 27-29 rows of ~9.4 MB therefore serialize into one ~272 MB response.
- The sender only logs when a message exceeds `rpc_max_message_size`: `src/yb/rpc/serialization.cc:74-77` is
  `LOG(DFATAL)`, a no-op in release builds, and the frame is written anyway.
- The receiving pggate client rejects the frame: `src/yb/rpc/binary_call_parser.cc:90-94` returns
  `NetworkError` "The frame had a length of $0, but we only support messages up to $1 bytes long" (the reporter
  cites `:94`). `src/yb/client/tablet_rpc.cc:259-261` treats a network error as retryable
  (`FailToNewReplica`), and `src/yb/rpc/rpc.cc:207-213` converts the exhausted deadline into the reported
  `Timed out: Read(tablet: ..., num_attempts: N, ...) passed its deadline ...: <last error>` message
  (format from `src/yb/client/async_rpc.cc:152-154`). `rpc_max_message_size` = 255_MB at
  `src/yb/rpc/yb_rpc.cc:44,104`; the 60 s deadline is `pggate_rpc_timeout_secs` (`pggate_flags.cc:29`).

Current code (master 52c3c2bf, identical mechanism on every resolved target, line numbers per tag in the matrix
file): `src/yb/docdb/pgsql_operation.cc:2822-2825` caps every DocDB read response at
`rpc_max_message_size x max_buffer_size_to_rpc_limit_ratio` (0.9, `src/yb/yql/pggate/pggate_flags.cc:198`) and
takes the client's `size_limit` when stricter; `:2882-2888` truncates the row that overflows the cap and returns a
paging state pointing at it (a row that alone exceeds the cap is still returned so the scan cannot loop);
`src/yb/yql/pggate/pg_doc_op.cc:369-370` divides that budget by the number of parallel requests when
`yb_fetch_size_limit` is 0 (its default, `src/yb/yql/pgwrapper/pg_wrapper.cc:292`; `yb_fetch_row_limit` default
1024 at `:289`). Senders now refuse oversized messages before writing them (`src/yb/rpc/serialization.cc:81-82,
114-115`, `src/yb/tserver/pg_client_session.cc:692-699`), so `binary_call_parser.cc:86-90` is no longer reachable
from a YSQL scan. Still affected targets: none of 2024.2 / 2025.1 / 2025.2 / 2026.1 / master. No master fix PR is
warranted for this issue.

## Trigger matrix
Harness: single-node yugabyted with `rpc_max_message_size=8388608` (so the failure needs ~8 MB pages instead of
255 MiB), 512 rows x 264,543-byte JSON array text in a 2-tablet table (~135 MB), reporter-shaped query (A), plain
full scan (B), one 7,700,000-byte value (C1, above the 0.9 x cap, below the limit), one 9,000,000-byte value
(C2, above the limit), `SET yb_fetch_size_limit='1MB'` (D), `SET yb_fetch_row_limit=8` (E), and a restart control:
era `--ysql_prefetch_limit=8` (thread workaround), current builds `--max_buffer_size_to_rpc_limit_ratio=1000`
(cap disabled, F). Predictions were written into the script header before execution.

| condition | expected | observed |
|---|---|---|
| era 2.1.6.0-b17, A/B default | reported signature after deadline | A: signature reproduced (118 attempts, 60.1 s; frame 66,402,851 B vs 8,388,608 limit); B: read timed out, same frame rejected in tserver log |
| era, C1 / C2 | OK / write fails | OK (7,700,001 B read back) / write timed out after 330 attempts, tserver rejected the inbound frame |
| era, restart with ysql_prefetch_limit=8 | OK | B 135,446,528 B, A 5 groups |
| 2024.2.11.0 / 2025.1.4.0 / 2025.2.6.0 / 2026.1.1.1 / 2.25.2.0, A/B default | OK | OK on all five (A 5 groups; B 135,446,528 B); signature text absent from every server log |
| same five, C1 | OK | OK (7,700,001 B read back) on all five |
| same five, C2 | fails without frame error | 2024.2 / 2025.1 / 2.25.2.0: sender refuses ("Sending too long RPC message"); 2025.2.6.0: client timed out on the shared-memory exchange; 2026.1.1.1: other non-frame error; row never stored on any build |
| same five, D / E | OK | OK on all five |
| same five, F cap disabled, B | sender refuses | refused on all five (page 66,402,811 B on 2024.2/2025.1/2.25.2.0; 135,451,152 B merged page on 2025.2/2026.1) |
| same five, F cap disabled + yb_fetch_size_limit=1MB, B | OK | OK 135,446,528 B on all five |
| 2.25.2.0 A/B default (prediction: merged parallel pages overflow) | FAIL | OK - falsified; that build does not merge the two tablet pages into one response (see matrix note) |

## Reproduction
Script: `/tmp/claude/yb4679/yb4679-matrix.sh <label> <image> [era]`; consolidated results and per-tag source
evidence: `/tmp/claude/yb4679/yb4679-matrix.md`. Images (all `docker.io/yugabytedb/yugabyte:`):
`2024.2.11.0-b36-aarch64`, `2025.1.4.0-b103-aarch64`, `2025.2.6.0-b111-aarch64`, `2026.1.1.1-b2-aarch64`,
`2.25.2.0-b359-aarch64`; era positive control `2.1.6.0-b17` (linux/amd64 under podman emulation).
Failure signature to grep for: `The frame had a length of`. Observed only on the era run (client error and tserver
WARNING log); absent on all five current targets. Control cases: D/E on every current build; the era workaround
restart; the cap-disabled restart on current builds shows the paging overflow is prevented by the cap and, when the
cap is removed, refused by the sender instead of surfacing as a retry storm.

## Versions tested
Resolved at: 2026-09-06T15:43Z by the controller and independently re-verified at 2026-09-06T15:57Z
Resolution evidence: Docker Hub `tags?name=<branch>&page_size=3&ordering=last_updated` per branch (87 `2024.2*`,
36 `2025.1*`, 39 `2025.2*`, 12 `2026.1*`, 9 `2.25.*`, 0 `2.27.`/`2.29.`/`2026.2` tags), `tags/latest` =
2026.1.1.1-b2 (manifest sha256:3926eedf...), GitHub `list_commits sha=<tag> perPage=1` for the tag SHAs, master HEAD
52c3c2bf (2026-09-05T19:58:56Z); the in-image `version_metadata.json` git hash matched the tag SHA on every run.

| target | latest resolved version | source revision | observed repro/control | effective status |
|---|---|---|---|---|
| 2024.2 | 2024.2.11.0-b36 (arm64 sha256:53c01c56315e245cc901cbc46e93f30fedb1f87833c7b1aa09d20fbeda27a288) | v2024.2.11.0 = 04544fd94fb4e50d05994784f444b075a4a5dfdb | A/B default OK; C1 OK; C2 refused; D/E OK; cap-disabled refused, +1MB OK | fixed |
| 2025.1 | 2025.1.4.0-b103 (sha256:08f1be0d6e074e0286e70067c12933a8cf536876d50fc43c52b1127d447ecdc0) | v2025.1.4.0 = 690ac4141e57d060fd834312f396643e8c41f260 | A/B default OK; C1 OK; C2 refused; D/E OK; cap-disabled refused, +1MB OK | fixed |
| 2025.2 | 2025.2.6.0-b111 (sha256:4a1f7a94d3b70c61ecaf5e049521d691b16c9bf204ed01bdb6ab75d5beeb623e) | v2025.2.6.0 = ae22ef35ff65de63303467d95ac3e6d2382feba9 | A/B default OK; C1 OK; C2 timed out without frame error; D/E OK; cap-disabled refused, +1MB OK | fixed |
| 2026.1 | 2026.1.1.1-b2 = latest (sha256:d27eede72bbc669e348e9920318589b4bb68e98204a241ec6bfaba04bf4f6699) | v2026.1.1.1 = 3b010e021f6027744bf9e016b17e108a2eb24e44 | A/B default OK; C1 OK; C2 non-frame error; D/E OK; cap-disabled refused, +1MB OK | fixed |
| master | preview 2.25.2.0-b359 (sha256:3f9cd1b8f520cc78e82ace51174be73343a551148bc6c9a9f821a22f1d2858c5); master HEAD source 52c3c2bf5800aad3e9ae74f4d41b224868caac09 | 03bbc9f31eae5c3fecabd8f0afa5e99211e6080d (in-image) | A/B default OK; C1 OK; C2 refused; D/E OK; cap-disabled refused, +1MB OK | fixed (runtime on the preview; master HEAD source carries the same cap plus the client split) |
| reported era | 2.1.6.0-b17 (linux/amd64, sha256:d36f215e874b40d3adb40d544c13cba13207e641d8ff8fdb452c5ed7682924d9) | 7eab087970603eadb0363fa747cc31e69e89dfc1 | A reproduces the signature; B fails on the same rejected frame; restart with ysql_prefetch_limit=8 makes A/B pass | affected (positive control) |

First-fixed releases (provenance from commit/tag ancestry, not runtime proof): hard DocDB cap 2115906c (#24627,
D39706, 2025-03-11) + 0.9 default 8922a47c (#26375, D42903, 2025-04-02); 2024.2 backports 2352f0bf (D42458) and
09f20747 (D42973) -> first in v2024.2.3.0 (absent through v2024.2.2.5); present from v2025.1.0.0, v2025.2.0.0,
v2026.1.0.0 and the v2.25.2.0 preview. Client-side per-request split abf33bac (#27202, D44455, 2025-06-10) -> first in
v2024.2.4.0, v2025.1.0.0, v2025.2.0.0, v2026.1.0.0; absent in all 2.25.x previews. GUCs `yb_fetch_size_limit` /
`yb_fetch_row_limit` from 99610b7a (#14392, D24658, 2023-04-28).

## Proposed fix direction
No code change needed for the reported defect. Optional hardening observed while testing (separate tickets if
pursued): (1) a write of a single value larger than `rpc_max_message_size` fails with a clear sender-side error on
2024.2/2025.1/2.25.2.0 but with a client timeout on 2025.2.6.0 (shared-memory exchange path) and an unrelated-looking
error on 2026.1.1.1 - a uniform "value exceeds rpc_max_message_size" error would be friendlier; (2) the issue thread
could be closed with a pointer to `yb_fetch_size_limit` / `yb_fetch_row_limit` and the 2025 default cap.

## Test strategy
Existing coverage: `PgMiniTest.ReadHugeRows` / `AnalyzeSmallRows` / `AnalyzeMediumRows` / `AnalyzeLargeRows`
(`src/yb/yql/pgwrapper/pg_mini-test.cc`, lowering `FLAGS_rpc_max_message_size`), `PgManyTabletsSelect.AnalyzeTableWithLargeRows`
(`pg_tablet_split-test.cc`, #27202), `org.yb.pgsql.TestPgRegressYbFetchLimits`. Public-entry-point check used here:
start yugabyted with `--rpc_max_message_size=8388608`, load wide rows, run the scan with default GUCs, then
`--max_buffer_size_to_rpc_limit_ratio=1000` as the negative control (`/tmp/claude/yb4679/yb4679-matrix.sh`).

## Confidentiality check
Synthetic schema (`t_json`, `t_one`), synthetic data (`label-N`, repeated characters), loopback addresses only,
official public images and public commit/tag identifiers; the reporter's public issue text is the only quoted
external material. Build hostnames from `version_metadata.json` were redacted in the logs. No customer identifiers,
PII, secrets, or internal URLs.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4679-matrix.md`):

# yugabyte/yugabyte-db#4679 (DB-2556) - runtime matrix and source evidence

Run date: 2026-09-06 (UTC 16:05-16:44). Host: Apple Silicon, podman 5.8.6 (applehv VM, arm64), official
`docker.io/yugabytedb/yugabyte` images; the era image is linux/amd64 and ran under podman's emulation.
Driver: `/tmp/claude/yb4679/yb4679-matrix.sh <label> <image> [era]` (one container per target, removed at the end).
Per-target raw output stayed under `/tmp/claude/yb4679/out/<label>/` (not attached; this file is the consolidated record).

## Harness (identical on every target)

- Single-node `yugabyted`, master and tserver started with `rpc_max_message_size=8388608` (8 MiB instead of the
  255 MiB default = 267386880) and `consensus_max_batch_size_bytes=2097152`; tserver also
  `pggate_rpc_timeout_secs=20`. Non-default tserver gflags are exported to PostgreSQL backends as `FLAGS_*`
  environment variables (`src/yb/yql/pgwrapper/pg_wrapper.cc:1123-1135` at v2024.2.11.0), so the same frame limit
  applies to the pggate client inside each backend.
- Data: `t_json (id int, jtl text, PRIMARY KEY (id HASH)) SPLIT INTO 2 TABLETS`, 512 rows, each `jtl` a 264,543-byte
  JSON array text (1050 elements; element index 1 is one of 5 labels, element index 8 is a bigint); 32 INSERT
  statements of 16 rows. `t_one (id int PRIMARY KEY, v text)` for the single-wide-row shapes.
- Cells:
  - `A_default`: reporter-shaped query `SELECT r->>1 AS label, avg(cast(r->>8 AS bigint)) FROM t_json tbl,
    json_array_elements(cast(tbl.jtl AS json)) r GROUP BY r->>1` with default GUCs (expects 5 groups).
  - `B_default`: `SELECT jtl FROM t_json` streamed to `wc -c` (expects ~135 MB).
  - `C1`: one 7,700,000-byte value (above the 0.9 x limit DocDB read cap of 7,549,747, below the 8,388,608 frame
    limit and the 0.95 x limit Raft single-op guard of 7,969,177 in `src/yb/tablet/preparer.cc`): INSERT then SELECT.
  - `C2`: one 9,000,000-byte value (above the frame limit): INSERT.
  - `D_size1MB`: `SET yb_fetch_size_limit = '1MB'` + query A. `E_row8`: `SET yb_fetch_row_limit = 8` + query A.
  - Era only: restart the tserver with `--ysql_prefetch_limit=8` (the workaround proposed in the issue thread), rerun B and A.
  - Modern only: restart the tserver with `--max_buffer_size_to_rpc_limit_ratio=1000` (disables the DocDB and client
    size caps), rerun B (`F_nocap_B`), then B with `SET yb_fetch_size_limit = '1MB'` (`F_nocap_size1MB_B`).
- Classification: `REPRO_FRAME` = the reported signature text "The frame had a length of" appears in the client error;
  `FAIL_TOO_LONG` = the sender refused with "Sending too long RPC message"; `FAIL_OTHER` = other error; `OK` = completed
  with the expected result. Server logs were grepped for the signature text at the end of every run.

## Predictions (recorded in the script header before the runs) vs observed

| target (image) | source rev | A_default | B_default | C1 insert/select | C2 insert | D_size1MB | E_row8 | restart control | signature in server logs |
|---|---|---|---|---|---|---|---|---|---|
| era 2.1.6.0-b17 (amd64) | 7eab0879 | pred REPRO_FRAME / **obs REPRO_FRAME** (118 attempts, 60.1 s deadline; frame 66,402,851 B vs 8,388,608 limit) | pred REPRO_FRAME / obs FAIL_OTHER (read RPC timed out; same 66,402,851 B frame rejected server-side) | pred OK / obs OK, 7,700,001 B read | pred fail / obs FAIL_OTHER (write timed out after 330 attempts; tserver rejected the inbound frame) | N/A (GUC absent) | N/A (GUC absent) | `--ysql_prefetch_limit=8`: B OK (135,446,528 B), A OK | yes (tserver WARNING log) |
| 2024.2.11.0-b36 | 04544fd9 | pred OK / obs OK | pred OK / obs OK 135,446,528 B | pred OK / obs OK, 7,700,001 B | pred FAIL_TOO_LONG / obs FAIL_TOO_LONG (9,000,227 B) | OK | OK | F_nocap_B FAIL_TOO_LONG (66,402,811 B page); F_nocap_size1MB_B OK | none |
| 2025.1.4.0-b103 | 690ac414 | pred OK / obs OK | pred OK / obs OK 135,446,528 B | pred OK / obs OK, 7,700,001 B | pred FAIL_TOO_LONG / obs FAIL_TOO_LONG (9,000,325 B) | OK | OK | F_nocap_B FAIL_TOO_LONG (66,402,811 B); F_nocap_size1MB_B OK | none |
| 2025.2.6.0-b111 | ae22ef35 | pred OK / obs OK | pred OK / obs OK 135,446,528 B | pred OK / obs OK, 7,700,001 B | pred FAIL_TOO_LONG / obs FAIL_OTHER (client timed out waiting on the shared-memory exchange; row not stored) | OK | OK | F_nocap_B FAIL_TOO_LONG (135,451,152 B combined); F_nocap_size1MB_B OK | none |
| 2026.1.1.1-b2 | 3b010e02 | pred OK / obs OK | pred OK / obs OK 135,446,528 B | pred OK / obs OK, 7,700,001 B | pred FAIL_TOO_LONG / obs FAIL_OTHER (non-frame error "Predecessor request ... was not applied"; row not stored) | OK | OK | F_nocap_B FAIL_TOO_LONG (135,451,152 B combined); F_nocap_size1MB_B OK | none |
| master preview 2.25.2.0-b359 | 03bbc9f3 | pred FAIL_TOO_LONG / **obs OK** (prediction falsified, see note) | pred FAIL_TOO_LONG / **obs OK** 135,446,528 B | pred OK / obs OK, 7,700,001 B | pred FAIL_TOO_LONG / obs FAIL_TOO_LONG (9,000,308 B) | OK | OK | F_nocap_B FAIL_TOO_LONG (66,402,811 B); F_nocap_size1MB_B OK | none |

Notes.
- Every 2024.2/2025.x/2026.1/preview run: 512 rows loaded, `length(jtl)` = 264,543 for id=1, query A returned its 5
  groups, query B streamed 135,446,528 bytes, and no run wrote the reported signature text to any server log.
- Era: query A failed with the reported shape (`Timed out: Read(tablet: ..., num_ops: 1, num_attempts: 118, ...)
  passed its deadline ... Network error (yb/rpc/binary_call_parser.cc:94): The frame had a length of 66402851, but
  we only support messages up to 8388608 bytes long`). The tserver WARNING log carries the same rejection
  (`connection.cc:290 ... Command sequence failure ...`). One 2-tablet page (256 rows x 264,543 B) is 66.4 MB, so the
  row-count-only paging of the era produces a frame eight times the configured limit, exactly the mechanism of the
  2020 report (327 rows x ~9.4 MB against the 255 MiB default).
- Preview 2.25.2.0 note: the prediction assumed the two parallel tablet pages (7.5 MB each) are merged into one
  response larger than 8 MiB. Observed: they are not merged on that build (the cap-disabled control shows a
  66,402,811-byte single-tablet page, the same as 2024.2/2025.1), so the default-GUC scan succeeds. 2025.2.6.0 and
  2026.1.1.1 do merge the two pages (135,451,152 bytes in the cap-disabled control) and still succeed under default
  GUCs because their client-side per-request cap divides the budget by the number of parallel requests.
- C2 (a single value larger than the frame limit) cannot be stored by any build; the newer builds refuse at the sender
  (`Sending too long RPC message`) or fail with another non-frame error, so the read-side residual "one row larger than
  rpc_max_message_size" is not reachable through YSQL writes made under the same limit. C1 shows that one row larger
  than the 0.9 x cap but smaller than the limit is written and read back intact on every build.
- The cap-disabled control (`max_buffer_size_to_rpc_limit_ratio=1000`) reproduces the 2020 paging behaviour (a
  66-135 MB page) on every current build, and every current build refuses to send it with a one-line error instead of
  emitting an oversized frame and retrying until the deadline.

## Source evidence per tag (files extracted with `git cat-file -p <tag>:<path>`)

| tag | DocDB default cap = rpc_max x ratio, min(client size_limit) | hard cap (overflowing row truncated, `fetched_rows > 0 &&`) | client per-send_count split | `yb_fetch_row_limit` / `yb_fetch_size_limit` defaults |
|---|---|---|---|---|
| v2.1.6 (era) | absent: `src/yb/docdb/pgsql_operation.cc:434-439,484` bounds the loop by `row_count_limit` only; pggate sets it from `ysql_prefetch_limit=1024` (`src/yb/yql/pggate/pggate_flags.cc:51`, `pg_doc_op.cc:106,295,314`) | absent | absent | GUCs absent |
| v2024.2.2.5 | soft only: `pgsql_operation.cc:2093-2101` ("just an effort to limit the damage") | absent | absent | 1024 / 0 |
| v2024.2.3.0 .. v2024.2.11.0 | `pgsql_operation.cc:2143-2146` (v2024.2.11.0) | `:2198` | from v2024.2.4.0 (`pg_doc_op.cc:402-403` at v2024.2.11.0) | 1024 / 0 (`pg_wrapper.cc:257,260`) |
| v2025.1.0.0 .. v2025.1.4.0 | `pgsql_operation.cc:2837-2840` (v2025.1.4.0) | `:2890` | `pg_doc_op.cc:392-393` | 1024 / 0 (`pg_wrapper.cc:275,278`) |
| v2025.2.0.0 .. v2025.2.6.0 | `pgsql_operation.cc:2788-2791` (v2025.2.6.0) | `:2844` | `pg_doc_op.cc:780-781` | 1024 / 0 (`pg_wrapper.cc:279,282`) |
| v2026.1.0.0 .. v2026.1.1.1 | `pgsql_operation.cc:2796-2799` (v2026.1.1.1) | `:2852` | `pg_doc_op.cc:363-364` | 1024 / 0 (`pg_wrapper.cc:288,291`) |
| v2.25.2.0 (preview) | `pgsql_operation.cc:2777-2780` | `:2830` | absent | 1024 / 0 (`pg_wrapper.cc:265,268`) |
| master 52c3c2bf (2026-09-05) | `pgsql_operation.cc:2822-2825` | `:2882-2888` | `pg_doc_op.cc:369-370` | 1024 / 0 (`pg_wrapper.cc:289,292`) |

Common to all builds: `src/yb/rpc/binary_call_parser.cc` rejects an inbound frame longer than `rpc_max_message_size`
(master `:86-90`; v2.1.6 `:90-94`, the line the reporter cites). Sender side: v2.1.6 `src/yb/rpc/serialization.cc:74-77`
only logs (DFATAL, a no-op in release builds) and sends anyway; current builds return `InvalidArgument` "Sending too
long RPC message" before sending (`serialization.cc:81-82,114-115`; `src/yb/tserver/pg_client_session.cc:692-699`).
`rpc_max_message_size` is 255 MiB and non-runtime on every build (`src/yb/util/pb_util.cc:128-143` on the release
tags, `:132-175` on master with the consensus validator); `max_buffer_size_to_rpc_limit_ratio` defaults to 0.9
(`src/yb/yql/pggate/pggate_flags.cc:198`).

## Fix provenance (commit ancestry; runtime proof is the table above)

- 99610b7a (2023-04-28, #14392, D24658): `yb_fetch_size_limit` GUC (default 0) and `yb_fetch_row_limit` GUC replacing
  the `ysql_prefetch_limit` gflag (deprecated `04_2023`).
- 2115906c (2025-03-11, #24627, D39706): `yb_fetch_size_limit` becomes a hard limit and `rpc_max_message_size` becomes
  the default DocDB response cap; the overflowing row is truncated and re-read on the next page (single row larger than
  the cap is still returned by design). 2024.2 backport 2352f0bf (2025-03-17, D42458).
- 8922a47c (2025-04-02, #26375, D42903): default cap lowered to `rpc_max_message_size x max_buffer_size_to_rpc_limit_ratio`
  (0.9) to leave room for headers. 2024.2 backport 09f20747 (2025-04-07, D42973).
- abf33bac (2025-06-10, #27202, D44455): client divides the size budget by the number of parallel requests.
- First release containing the hard cap + 0.9 default, by tag content: 2024.2 -> v2024.2.3.0 (absent through
  v2024.2.2.5); 2025.1 -> v2025.1.0.0; 2025.2 -> v2025.2.0.0; 2026.1 -> v2026.1.0.0; preview -> v2.25.2.0. The client
  split: 2024.2 -> v2024.2.4.0; 2025.1/2025.2/2026.1 -> first release of each branch; absent in every 2.25.x preview.

## Target resolution

Resolved 2026-09-06T15:43Z by the controller and independently re-verified at 2026-09-06T15:57Z (Docker Hub
`tags?name=<branch>&page_size=3&ordering=last_updated`, `tags/latest`, and GitHub `list_commits sha=<tag> perPage=1`):
2024.2.11.0-b36 (arm64 sha256:53c01c56...; newest of 87 `2024.2*` tags), 2025.1.4.0-b103 (sha256:08f1be0d...; 36),
2025.2.6.0-b111 (sha256:4a1f7a94...; 39), 2026.1.1.1-b2 = `latest` (sha256:d27eede7...; manifest sha256:3926eedf...; 12),
2.25.2.0-b359 (sha256:3f9cd1b8...; 9; `2.27.`/`2.29.`/`2026.2` counts 0). Tag SHAs: v2024.2.11.0 = 04544fd9,
v2025.1.4.0 = 690ac414, v2025.2.6.0 = ae22ef35, v2026.1.1.1 = 3b010e02, v2.25.2.0 = 03bbc9f3; master HEAD 52c3c2bf
(2026-09-05T19:58:56Z). The in-image `version_metadata.json` git hashes matched these SHAs on every run.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4679-matrix.sh.txt`):

#!/usr/bin/env bash
# Runtime matrix for yugabyte/yugabyte-db#4679 (DB-2556):
#   "The frame had a length of 272035850, but we only support messages up to 267386880 bytes long"
#
# Mechanism under test: a YSQL sequential scan is paged by ROW COUNT only (ysql_prefetch_limit /
# yb_fetch_row_limit = 1024), so a page of wide rows can exceed rpc_max_message_size (255 MiB
# default); the receiving side's BinaryCallParser rejects the frame and the client retries until
# its deadline. To make this cheap and deterministic we LOWER the frame limit to 8 MiB
# (rpc_max_message_size=8388608) and load 512 rows x ~256 KB (~134 MB) into a 2-tablet table.
#
# Usage: yb4679-matrix.sh <label> <image> [era]
#   e.g. yb4679-matrix.sh 2025_2 docker.io/yugabytedb/yugabyte:2025.2.6.0-b111-aarch64
#        yb4679-matrix.sh era docker.io/yugabytedb/yugabyte:2.1.6.0-b17 era   (linux/amd64 emulation)
#
# PREDICTIONS (stated before running):
#   era (2.1.6.0-b17, row-count paging only, sender only LOG(DFATAL)s oversized messages):
#     A_default/B_default -> REPRO_FRAME (the reported signature, after the RPC deadline)
#     C1 (single 7,700,000 B row: above the 0.9*limit read cap, below the frame limit) -> OK
#     C2 (single 9,000,000 B row > limit) -> INSERT fails (tserver rejects the inbound frame)
#     D/E (yb_fetch_size_limit / yb_fetch_row_limit GUCs) -> N/A (GUCs do not exist)
#     E_restart (tserver --ysql_prefetch_limit=8, the workaround suggested in the thread) -> OK
#   2024.2.11.0 / 2025.1.4.0 / 2025.2.6.0 / 2026.1.1.1 (DocDB hard cap 0.9*rpc_max + client split):
#     A_default/B_default -> OK ; C1 -> OK ; C2 -> FAIL_TOO_LONG at INSERT ("Sending too long RPC
#     message", no frame error, no retry storm) ; D/E -> OK ;
#     F_nocap (restart with max_buffer_size_to_rpc_limit_ratio=1000, i.e. cap disabled) -> FAIL_TOO_LONG
#     F_size1MB (cap disabled but SET yb_fetch_size_limit='1MB') -> OK
#   2.25.2.0-b359 (DocDB hard cap present, client-side per-send_count split #27202 ABSENT):
#     A_default/B_default -> FAIL_TOO_LONG (two parallel 7.5 MB tablet pages combined into one
#     Perform response > 8 MiB) ; D/E -> OK ; C1 -> OK ; C2 -> FAIL_TOO_LONG ; F -> FAIL_TOO_LONG
#   The literal signature "The frame had a length of" is predicted on the era cell ONLY.
set -u
LABEL="${1:?label}"; IMAGE="${2:?image}"; MODE="${3:-modern}"
C="yb4679-${LABEL}"
OUT="/tmp/claude/yb4679/out/${LABEL}"; mkdir -p "$OUT"
LOG="$OUT/run.log"; : > "$LOG"
log() { printf '%s\n' "$*" | tee -a "$LOG"; }
verdict() { log ">>> $LABEL $1"; }

RPCMAX=8388608          # 8 MiB frame limit (default is 255 MiB = 267386880)
CONS=2097152            # consensus_max_batch_size_bytes must stay < rpc_max_message_size - 1KB
COMMON="rpc_max_message_size=${RPCMAX},consensus_max_batch_size_bytes=${CONS}"
TS_BASE="${COMMON},pggate_rpc_timeout_secs=20"   # shorten the client retry deadline (60s default)

PLATFORM=""
if [ "$MODE" = "era" ]; then PLATFORM="--platform linux/amd64"; HOST=127.0.0.1; else HOST=""; fi

start_cluster() { # $1 = extra tserver flags (comma-separated, may be empty)
  local extra="$1" ts="$TS_BASE"
  [ -n "$extra" ] && ts="${ts},${extra}"
  if [ "$MODE" = "era" ]; then
    podman exec "$C" bash -c "/home/yugabyte/bin/yugabyted start --bind_ip=127.0.0.1 --ui=false --base_dir=/tmp/ybd --master_flags='${COMMON}' --tserver_flags='${ts}'" >>"$OUT/yugabyted.log" 2>&1
  else
    podman exec "$C" bash -c "/home/yugabyte/bin/yugabyted start --advertise_address=\$(hostname -i) --base_dir=/tmp/ybd --master_flags='${COMMON}' --tserver_flags='${ts}'" >>"$OUT/yugabyted.log" 2>&1
  fi
}
stop_cluster() { podman exec "$C" bash -c "/home/yugabyte/bin/yugabyted stop --base_dir=/tmp/ybd" >>"$OUT/yugabyted.log" 2>&1; sleep 3; }

sql() {  # sql <file-tag> <sql...>  -> runs via ysqlsh, stdout+stderr to $OUT/<tag>.txt, echoes file
  local tag="$1"; shift
  podman exec -i "$C" bash -c "/home/yugabyte/bin/ysqlsh -h \${YBH:-\$(hostname -i)} -U yugabyte -d yugabyte -v ON_ERROR_STOP=1 -X -q" <<<"$*" >"$OUT/$tag.txt" 2>&1
  echo "$OUT/$tag.txt"
}
sql_bytes() {  # sql_bytes <tag> <sql> -> stdout byte count into $OUT/<tag>.bytes, stderr into $OUT/<tag>.txt
  local tag="$1"; shift
  podman exec -i "$C" bash -c "/home/yugabyte/bin/ysqlsh -h \${YBH:-\$(hostname -i)} -U yugabyte -d yugabyte -v ON_ERROR_STOP=1 -X -q -tA 2>/tmp/err.txt | wc -c; cat /tmp/err.txt >&2" <<<"$*" >"$OUT/$tag.bytes" 2>"$OUT/$tag.txt"
}
classify() {  # classify <file> [ok-predicate-file-or-empty]
  local f="$1"
  if grep -q 'The frame had a length of' "$f"; then echo REPRO_FRAME
  elif grep -q 'Sending too long RPC message\|Sending too long of an RPC message' "$f"; then echo FAIL_TOO_LONG
  elif grep -q -E '^(ERROR|FATAL|psql|ysqlsh):|^ERROR|Timed out|Network error|could not connect' "$f"; then echo FAIL_OTHER
  else echo OK; fi
}
firsterr() { grep -m1 -E 'ERROR|FATAL|Timed out|Network error|frame had|too long' "$1" | cut -c1-260; }
frame_in_server_logs() { podman exec "$C" bash -c "grep -rl 'The frame had a length of' /tmp/ybd 2>/dev/null | head -3; grep -rh -m1 'The frame had a length of' /tmp/ybd 2>/dev/null | head -1 | cut -c1-300"; }

wait_ready() {
  local i
  for i in $(seq 1 200); do
    if podman exec "$C" bash -c "/home/yugabyte/bin/ysqlsh -h \${YBH:-\$(hostname -i)} -U yugabyte -d yugabyte -Atqc 'select 1'" 2>/dev/null | grep -qx 1; then log "### ysql ready after ~$((i*3))s"; return 0; fi
    sleep 3
  done
  log "### ysql NOT ready after 600s"; return 1
}

# ---------------------------------------------------------------- start
podman rm -f "$C" >/dev/null 2>&1 || true
log "### $(date -u +%FT%TZ) label=$LABEL image=$IMAGE mode=$MODE"
log "### image digest: $(podman image inspect --format '{{.Digest}}' "$IMAGE" 2>/dev/null)"
if [ "$MODE" = "era" ]; then
  podman run -d --name "$C" $PLATFORM -e YBH=127.0.0.1 "$IMAGE" bash -c 'tail -f /dev/null' >/dev/null
else
  podman run -d --name "$C" "$IMAGE" bash -c 'tail -f /dev/null' >/dev/null
fi
log "### version_metadata.json: $(podman exec "$C" bash -c 'tr -d "\n " < /home/yugabyte/version_metadata.json 2>/dev/null' | sed -E 's/"build_hostname":"[^"]*"/"build_hostname":"<redacted>"/')"
log "### flags: master[$COMMON] tserver[$TS_BASE]"
start_cluster ""
wait_ready || { verdict "STARTUP_FAILED"; podman logs "$C" >>"$LOG" 2>&1; exit 1; }
log "### version(): $(podman exec "$C" bash -c "/home/yugabyte/bin/ysqlsh -h \${YBH:-\$(hostname -i)} -U yugabyte -d yugabyte -Atqc 'select version()'" 2>&1)"
tsflags() { podman exec "$C" bash -c 'for p in $(pgrep -f "bin/yb-tserver"); do tr "\0" "\n" < /proc/$p/cmdline; done 2>/dev/null | grep -E "^--(rpc_max_message_size|max_buffer_size_to_rpc_limit_ratio|ysql_prefetch_limit|pggate_rpc_timeout_secs|consensus_max_batch_size_bytes)=" | sort -u | tr "\n" " "'; }
srvlog() { podman exec "$C" bash -c "grep -rh -m2 -E '$1' /tmp/ybd 2>/dev/null | grep -v ysqlsh | head -${2:-2} | cut -c1-320"; }
log "### tserver cmdline flags: $(tsflags)"

# ---------------------------------------------------------------- data
# t_json: 512 rows, each a ~256 KB JSON array text (1050 elements of ~250 B; element index 1 is a
# label with 5 distinct values, element index 8 is a bigint), 2 tablets => ~67 MB per tablet.
f=$(sql create_t_json "CREATE TABLE t_json (id int, jtl text, PRIMARY KEY (id HASH)) SPLIT INTO 2 TABLETS;")
if grep -q ERROR "$f"; then
  log "### SPLIT INTO unsupported/failed ($(firsterr "$f")); creating without SPLIT"
  f=$(sql create_t_json "CREATE TABLE t_json (id int PRIMARY KEY, jtl text);")
fi
grep -q ERROR "$f" && { verdict "SETUP_FAILED create table: $(firsterr "$f")"; exit 1; }
ROWGEN="(SELECT json_agg(json_build_array('x', 'label-' || (g % 5), repeat('p', 200), 'b', 'c', 'd', 'e', 'f', g)) FROM generate_series(1, 1050) g)::text"
for chunk in $(seq 0 31); do
  a=$((chunk*16+1)); b=$((chunk*16+16))
  f=$(sql "load_$chunk" "WITH j AS (SELECT $ROWGEN AS jtl) INSERT INTO t_json SELECT i, j.jtl FROM j, generate_series($a, $b) AS i;")
  grep -q ERROR "$f" && { verdict "SETUP_FAILED load chunk $chunk: $(firsterr "$f")"; break; }
done
f=$(sql rowcount "SELECT count(*) FROM t_json;"); log "### t_json rows: $(grep -E '^\s*[0-9]+' "$f" | tr -d ' ')"
f=$(sql rowlen "SELECT length(jtl) FROM t_json WHERE id = 1;"); log "### t_json length(jtl) of id=1: $(grep -E '^\s*[0-9]+' "$f" | tr -d ' ') bytes"

QA="SELECT r->>1 AS label, avg(cast(r->>8 AS bigint)) AS avgtime FROM t_json tbl, json_array_elements(cast(tbl.jtl AS json)) r GROUP BY r->>1 ORDER BY 1;"
QB="SELECT jtl FROM t_json;"

run_A() { # run_A <tag> <prefix-sql>
  local f; f=$(sql "$1" "$2 $QA"); local r; r=$(classify "$f")
  if [ "$r" = OK ]; then local n; n=$(grep -c 'label-' "$f"); [ "$n" = 5 ] || r="OK_BUT_ROWS=$n"; fi
  verdict "$1 $r $( [ "$r" != OK ] && firsterr "$f")"
}
run_B() { # run_B <tag> <prefix-sql>
  sql_bytes "$1" "$2 $QB"; local r; r=$(classify "$OUT/$1.txt"); local by; by=$(tr -d ' \n' < "$OUT/$1.bytes")
  if [ "$r" = OK ]; then [ "${by:-0}" -gt 100000000 ] 2>/dev/null || r="OK_BUT_BYTES=$by"; fi
  verdict "$1 $r bytes=$by $( [ "$r" != OK ] && firsterr "$OUT/$1.txt")"
}

# ---------------------------------------------------------------- cells
run_A A_default ""
run_B B_default ""
# C: single wide rows. C1 (7,700,000 B) sits above the 0.9*limit DocDB read cap (7,549,747) but
#    below both the frame limit (8,388,608) and the Raft leader single-op guard
#    (0.95*limit = 7,969,177; tablet/preparer.cc). C2 (9,000,000 B) exceeds the frame limit itself.
f=$(sql c_create "CREATE TABLE t_one (id int PRIMARY KEY, v text);")
f=$(sql C1_insert "INSERT INTO t_one VALUES (1, repeat('y', 7700000));"); r=$(classify "$f"); verdict "C1_insert_7700000B $r $( [ "$r" != OK ] && firsterr "$f")"
sql_bytes C1_select "SELECT v FROM t_one WHERE id = 1;"; r=$(classify "$OUT/C1_select.txt"); verdict "C1_select_7700000B $r bytes=$(tr -d ' \n' < "$OUT/C1_select.bytes") $( [ "$r" != OK ] && firsterr "$OUT/C1_select.txt")"
f=$(sql C2_insert "INSERT INTO t_one VALUES (2, repeat('z', 9000000));"); r=$(classify "$f"); verdict "C2_insert_9000000B $r $( [ "$r" != OK ] && firsterr "$f")"
log "### server log lines for C2 (9000000-byte op): $(srvlog 'too long|exceeds limit|frame had a length|9000[0-9]{3}' 3 | tr '\n' ' | ')"
if [ "$r" = OK ]; then sql_bytes C2_select "SELECT v FROM t_one WHERE id = 2;"; r=$(classify "$OUT/C2_select.txt"); verdict "C2_select_9000000B $r bytes=$(tr -d ' \n' < "$OUT/C2_select.bytes") $( [ "$r" != OK ] && firsterr "$OUT/C2_select.txt")"; fi
f=$(sql c_count "SELECT count(*) FROM t_one;"); log "### t_one rows after C1/C2: $(grep -E '^\s*[0-9]+' "$f" | tr -d ' ')"
# D/E: user-level controls (GUCs introduced 2023-04, #14392)
run_A D_size1MB "SET yb_fetch_size_limit = '1MB';"
run_A E_row8 "SET yb_fetch_row_limit = 8;"
log "### server-side frame rejections so far: $(frame_in_server_logs | tr '\n' ' ')"

# ---------------------------------------------------------------- restart-based controls
if [ "$MODE" = "era" ]; then
  log "### restarting tserver with ysql_prefetch_limit=8 (thread workaround)"
  stop_cluster; start_cluster "ysql_prefetch_limit=8"; wait_ready || verdict "E_restart STARTUP_FAILED"
  log "### tserver cmdline flags after restart: $(tsflags)"
  run_B E_restart_prefetch8_B ""
  run_A E_restart_prefetch8_A ""
else
  log "### restarting tserver with max_buffer_size_to_rpc_limit_ratio=1000 (DocDB/client size cap effectively disabled)"
  stop_cluster; start_cluster "max_buffer_size_to_rpc_limit_ratio=1000"; wait_ready || verdict "F STARTUP_FAILED"
  log "### tserver cmdline flags after restart: $(tsflags)"
  run_B F_nocap_B ""
  run_B F_nocap_size1MB_B "SET yb_fetch_size_limit = '1MB';"
fi
log "### server-side frame rejections at end: $(frame_in_server_logs | tr '\n' ' ')"
log "### $(date -u +%FT%TZ) done $LABEL"
grep '^>>>' "$LOG" > "$OUT/summary.txt"
podman rm -f "$C" >/dev/null 2>&1 || true

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
